crates/noirc_evaluator/src/ssa/function.rs

```diff
@@ -0,0 +1,208 @@
+use noirc_frontend::node_interner::FuncId;
+
+use super::{
+    code_gen::IRGenerator,
+    context::SsaContext,
+    node::{self, NodeId, ObjectType},
+    ssa_form,
+};
+
+pub struct SSAFunction<'a> {
+    pub igen: IRGenerator<'a>,
+    pub id: FuncId,
+    //signature..
+    pub arguments: Vec<NodeId>,
+}
+
+impl<'a> SSAFunction<'a> {
+    //Parse the AST function body into ssa form in cfg
+    pub fn parse_statements(
+        &mut self,
+        block: &[noirc_frontend::node_interner::StmtId],
+        env: &mut Environment,
+    ) {
+        for stmt in block {
+            self.igen.evaluate_statement(env, stmt).unwrap();
+        }
+    }
+
+    pub fn new(func: FuncId, ctx: &'a noirc_frontend::hir::Context) -> SSAFunction<'a> {
+        SSAFunction {
+            igen: IRGenerator::new(ctx),
+            id: func,
+            arguments: Vec::new(),
+        }
+    }
+
+    pub fn compile(&mut self) -> Option<NodeId> {
+        //Optimisation
+        super::block::compute_dom(&mut self.igen.context);
+        super::optim::cse(&mut self.igen.context);
+        //Unrolling
+        super::flatten::unroll_tree(&mut self.igen.context);
+        super::optim::cse(&mut self.igen.context)
+    }
+
+    //generates an instruction for calling the function
+    pub fn call(
+        func: FuncId,
+        arguments: &[noirc_frontend::node_interner::ExprId],
+        eval: &mut IRGenerator,
+        env: &mut Environment,
+    ) -> NodeId {
+        let call_id = eval.context.new_instruction(
+            NodeId::dummy(),
+            NodeId::dummy(),
+            node::Operation::Call(func),
+            node::ObjectType::NotAnObject, //TODO how to get the function return type?
+        );
+        let ins_arguments = eval.expression_list_to_objects(env, arguments);
+        let call_ins = eval.context.get_mut_instruction(call_id);
+        call_ins.ins_arguments = ins_arguments;
+        call_id
+    }
+
+    pub fn get_mapped_value(
+        func_id: noirc_frontend::node_interner::FuncId,
+        var: Option<&NodeId>,
+        ctx: &mut SsaContext,
+        inline_map: &HashMap<NodeId, NodeId>,
+    ) -> NodeId {
+        if let Some(&node_id) = var {
+            if node_id == NodeId::dummy() {
+                return node_id;
+            }
+            let mut my_const = None;
+            let node_obj_opt = ctx.functions_cfg[&func_id]
+                .igen
+                .context
+                .try_get_node(node_id);
+            if let Some(node::NodeObj::Const(c)) = node_obj_opt {
+                my_const = Some((c.get_value_field(), c.value_type));
+            }
+            if let Some(c) = my_const {
+                ctx.get_or_create_const(c.0, c.1)
+            } else {
+                dbg!(&node_id);
+                *inline_map.get(&node_id).unwrap()
+            }
+        } else {
+            NodeId::dummy()
+        }
+    }
+}
+
+pub fn get_result_type(op: OPCODE) -> (u32, ObjectType) {
```

Comment:
What is the u32 here? Looks like bitcount but then I'm not sure what all the 1's and 2's mean below.

Reply:
It corresponds to the number of elements returned. I have added a comment to explain this.